crates/bevy_ecs/src/bundle.rs

```diff
@@ -572,192 +502,368 @@ impl BundleInfo {
-                    location.table_row,
-                    self.change_tick,
-                    bundle,
+                    bundle_info
+                        .iter_components()
+                        .zip(add_bundle.bundle_status.iter())
+                        .filter(|(_, &status)| status == ComponentStatus::Added)
+                        .map(|(id, _)| id),
                 );
+            }
+            if archetype.has_on_insert() {
+                world.trigger_on_insert(entity, bundle_info.iter_components());
+            }
+        };
+        match &mut self.result {
+            InsertBundleResult::SameArchetype => {
+                {
+                    // SAFETY: Mutable references do not alias and will be dropped after this block
+                    let sparse_sets = {
+                        let world = self.world.world_mut();
+                        &mut world.storages.sparse_sets
+                    };
+                    let table = &mut *self.table;
+                    bundle_info.write_components(
+                        table,
+                        sparse_sets,
+                        add_bundle,
+                        entity,
+                        location.table_row,
+                        self.change_tick,
+                        bundle,
+                    );
+                }
+                // SAFETY: We have no oustanding mutable references to world as they were dropped
+                trigger_hooks(archetype, self.world.into_deferred());
                 location
             }
             InsertBundleResult::NewArchetypeSameTable { new_archetype } => {
-                let result = self.archetype.swap_remove(location.archetype_row);
-                if let Some(swapped_entity) = result.swapped_entity {
-                    let swapped_location =
-                        // SAFETY: If the swap was successful, swapped_entity must be valid.
-                        unsafe { self.entities.get(swapped_entity).debug_checked_unwrap() };
-                    self.entities.set(
-                        swapped_entity.index(),
-                        EntityLocation {
-                            archetype_id: swapped_location.archetype_id,
-                            archetype_row: location.archetype_row,
-                            table_id: swapped_location.table_id,
-                            table_row: swapped_location.table_row,
-                        },
+                let new_location = {
+                    // SAFETY: Mutable references do not alias and will be dropped after this block
+                    let table = &mut *self.table;
+                    let archetype = &mut *self.archetype;
+                    let new_archetype = &mut **new_archetype;
+                    let (sparse_sets, entities) = {
+                        let world = self.world.world_mut();
+                        (&mut world.storages.sparse_sets, &mut world.entities)
+                    };
```

Comment:
I am open to refactoring to be based on `UnsafeWorldCell` as I agree it's safer, however `table`, `archetype` and `new_archetype` aren't actually references to the fields of the `world` struct directly as there is indirection through `Vec`.

Reply:
Managed to avoid these situations by turning the pointers into references only after calls to `world_mut()` to ensure there are no outstanding references when that happens.